### PR TITLE
fix gas used and fee relayed v3

### PR DESCRIPTION
**Pull request by miiu96**
## Reasoning behind the pull request
- Fixed how is calculated `gasUsed` and `fee` fields for `relayed v3` transactions  
## Proposed changes
- Implemented new functions that will calculate the gas used and fee filed for the API transactions and for the indexer transactions
- Changed the version of the `mx-chain-core-go` : https://github.com/multiversx/mx-chain-core-go/pull/330
- Changed the version of the `mx-chain-es-indexer-go`: https://github.com/multiversx/mx-chain-es-indexer-go/pull/298
## Testing procedure
- Do a system test and check if the gas used and the fee is set correctly for the API transactions and the indexed transactions

## Pre-requisites

Based on the [Contributing Guidelines](https://github.com/multiversx/mx-chain-go/blob/master/.github/CONTRIBUTING.md#branches-management) the PR author and the reviewers must check the following requirements are met:
- was the PR targeted to the correct branch?
- if this is a larger feature that probably needs more than one PR, is there a `feat` branch created?
- if this is a `feat` branch merging, do all satellite projects have a proper tag inside `go.mod`?

### CODE REVIEW

File: process/economics/economicsData.go [hunk 1]
```diff
@@ -580,6 +591,16 @@ func (ed *economicsData) ComputeGasUsedAndFeeBasedOnRefundValue(tx data.Transact
 	return ed.ComputeGasUsedAndFeeBasedOnRefundValueInEpoch(tx, refundValue, currentEpoch)
 }
 
+// ComputeGasUnitsFromRefundValue will compute the gas unit based on the refund value
+func (ed *economicsData) ComputeGasUnitsFromRefundValue(tx data.TransactionWithFeeHandler, refundValue *big.Int) uint64 {
+	currentEpoch := ed.enableEpochsHandler.GetCurrentEpoch()
```
Comment (sstanculeanu) [inline process/economics/economicsData.go:596]: epoch should be the one of tx here as well, so perhaps get it as parameter

File: outport/process/transactionsfee/transactionsFeeProcessor_test.go [hunk 2]
```diff
@@ -597,3 +601,108 @@ func TestMoveBalanceWithSignalError(t *testing.T) {
...
+	arg.TxFeeCalculator = newEconomicsData(t)
+	txsFeeProc, err := NewTransactionsFeeProcessor(arg)
+	require.NotNil(t, txsFeeProc)
+	require.Nil(t, err)
+
+	err = txsFeeProc.PutFeeAndGasUsed(pool, 0)
+	require.Nil(t, err)
+	require.Equal(t, uint64(180_150_000), initialTx.GetFeeInfo().GetGasUsed())
+	require.Equal(t, "1074000000000000", initialTx.GetFeeInfo().GetFee().String())
+	require.Equal(t, "1074000000000000", initialTx.GetFeeInfo().GetInitialPaidFee().String())
```
Comment (sstanculeanu) [inline outport/process/transactionsfee/transactionsFeeProcessor_test.go:707]: shouldn't these values be the same as in TestComputeAndAttachGasUsedAndFeeRelayedV3WithAllInnerTxFailed:
```go
uint64(160766000),
"2670920000000000"
"2864760000000000"
```

File: testscommon/feeComputerStub.go [hunk 3]
```diff
@@ -13,6 +13,15 @@ type FeeComputerStub struct {
 	ComputeTxFeeBasedOnGasUsedCalled             func(tx *transaction.ApiTransactionResult, gasUsed uint64) *big.Int
 	ComputeGasLimitCalled                        func(tx *transaction.ApiTransactionResult) uint64
 	ComputeMoveBalanceFeeCalled                  func(tx *transaction.ApiTransactionResult) *big.Int
+	ComputeGasUnitForRelayedV3Called             func(tx *transaction.ApiTransactionResult) uint64
+}
+
+func (stub *FeeComputerStub) ComputeGasUnitForRelayedV3(tx *transaction.ApiTransactionResult) uint64 {
```
Comment (sstanculeanu) [inline testscommon/feeComputerStub.go:19]: missing comment

File: outport/process/transactionsfee/transactionsFeeProcessor.go [hunk 4]
```diff
@@ -258,6 +258,22 @@ func (tep *transactionsFeeProcessor) prepareRelayedTxV3WithResults(txHashHex str
 		refundsValue.Add(refundsValue, scr.Value)
 	}
 
+	isRelayedV3 := len(txWithResults.GetTxHandler().GetUserTransactions()) > 0
```
Comment (sstanculeanu) [inline outport/process/transactionsfee/transactionsFeeProcessor.go:261]: should also check if it has refunds here?

File: testscommon/feeComputerStub.go (hunk 3 above)
Comment (miiu96) [reply testscommon/feeComputerStub.go:19]: added

File: outport/process/transactionsfee/transactionsFeeProcessor.go (hunk 4 above)
Comment (miiu96) [reply outport/process/transactionsfee/transactionsFeeProcessor.go:261]: In this case no, the refunds are in the Smart Contract Results map.

File: process/economics/economicsData.go (hunk 1 above)
Comment (miiu96) [reply process/economics/economicsData.go:596]: changed

**Review by sstanculeanu — APPROVED** [review]

File: outport/process/transactionsfee/transactionsFeeProcessor.go [hunk 5]
```diff
@@ -328,6 +344,17 @@ func (tep *transactionsFeeProcessor) prepareScrsNoTx(transactionsAndScrs *transa
 			continue
 		}
 
+		isRelayedV3 := len(txFromStorage.InnerTransactions) > 0
+		if isRelayedV3 {
+			gasUnits := tep.txFeeCalculator.ComputeGasUnitsFromRefundValue(txFromStorage, scr.Value, epoch)
+
+			scrHandler.GetFeeInfo().SetForRelayed()
+			scrHandler.GetFeeInfo().SetGasUsed(gasUnits)
```
Comment (AdoAdoAdo) [inline outport/process/transactionsfee/transactionsFeeProcessor.go:352]: can you add a new field for gasUnitsRefunded

File: testscommon/economicsmocks/economicsHandlerMock.go [hunk 6]
```diff
@@ -52,6 +52,16 @@ type EconomicsHandlerMock struct {
 	SetTxTypeHandlerCalled                              func(txTypeHandler process.TxTypeHandler) error
 }
 
+// ComputeRelayedTxV3GasUnits -
+func (ehm *EconomicsHandlerMock) ComputeRelayedTxV3GasUnits(_ data.TransactionWithFeeHandler, _ uint32) uint64 {
```
[the remark this answers is not in the document]
Comment (miiu96) [reply testscommon/economicsmocks/economicsHandlerMock.go:56]: done

File: testscommon/economicsmocks/economicsDataHandlerStub.go [hunk 7]
```diff
@@ -52,6 +52,16 @@ type EconomicsHandlerStub struct {
 	ComputeMoveBalanceFeeInEpochCalled                  func(tx data.TransactionWithFeeHandler, epoch uint32) *big.Int
 }
 
+// ComputeRelayedTxV3GasUnits -
+func (e *EconomicsHandlerStub) ComputeRelayedTxV3GasUnits(_ data.TransactionWithFeeHandler, _ uint32) uint64 {
```
[the remark this answers is not in the document]
Comment (miiu96) [reply testscommon/economicsmocks/economicsDataHandlerStub.go:56]: done

File: outport/process/transactionsfee/transactionsFeeProcessor.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (miiu96) [reply outport/process/transactionsfee/transactionsFeeProcessor.go:261]: refactored

File: outport/process/transactionsfee/transactionsFeeProcessor.go (hunk 5 above)
Comment (miiu96) [reply outport/process/transactionsfee/transactionsFeeProcessor.go:352]: refactored

**Pull request closed without merging**